packages/server/src/core/bank_rule/specialization_groups.rs

```diff
@@ -1,71 +1,197 @@
+                }
+                if !completed_current_demand {
+                    return false;
+                }
+            }
+        }
+    }
+    true
+}
+
+// This function is looking for a valid assignment for the courses which fulfill the sgs requirements
+// If an assignment is found it returns it, None otherwise.
+fn find_valid_assignment_for_courses(
+    sgs: &[SpecializationGroup],
+    groups_indices: &[u8],
+    optional_sgs_for_course: &HashMap<CourseId, Vec<u8>>, // list of all optional sgs for each course
+    course_id_to_sg_index: &mut HashMap<CourseId, u8>,
+    course_index: usize, // course_index-th element in optional_sgs_for_course
+) -> Option<HashMap<CourseId, u8>> {
+    if course_index >= optional_sgs_for_course.len() {
+        if check_courses_assignment_for_sgs(sgs, groups_indices, course_id_to_sg_index) {
+            return Some(course_id_to_sg_index.clone());
+        }
+        return None;
+    }
+    if let Some((course_id, optional_groups)) = optional_sgs_for_course.iter().nth(course_index) {
+        for sg_index in optional_groups {
+            course_id_to_sg_index.insert(course_id.clone(), *sg_index);
+            if let Some(valid_assignment) = find_valid_assignment_for_courses(
+                sgs,
+                groups_indices,
+                optional_sgs_for_course,
+                course_id_to_sg_index,
+                course_index + 1,
+            ) {
+                return Some(valid_assignment);
+            }
+        }
+    }
+    None
+}
+
+fn check_if_completed_groups(
+    sgs: &[SpecializationGroup],
+    groups_indices: &[u8],
+    courses: &[CourseId],
+) -> Option<HashMap<CourseId, u8>> {
+    let mut optional_sgs_for_course = HashMap::<CourseId, Vec<u8>>::new();
+    for course_id in courses {
+        let mut relevant_groups_for_course = Vec::new();
+        for sg_index in groups_indices {
+            if sgs[*sg_index as usize].course_list.contains(course_id) {
+                relevant_groups_for_course.push(*sg_index);
+            }
+        }
+        if !relevant_groups_for_course.is_empty() {
+            // only this subset specialization groups consist course_id
+            optional_sgs_for_course.insert(course_id.clone(), relevant_groups_for_course);
+        }
+    }
+
+    let mut courses_assignment = HashMap::new();
+    find_valid_assignment_for_courses(
+        sgs,
+        groups_indices,
+        &optional_sgs_for_course,
+        &mut courses_assignment,
+        0,
+    )
+}
+
+// generates all subsets of size specialization_groups.groups_number and checks if one of them is fulfilled
+fn generate_subsets(
+    sgs: &[SpecializationGroup],
+    required_number_of_groups: u8,
+    sg_index: u8,
+    groups_indices: &mut Vec<u8>,
+    courses: &[CourseId],
+) -> Option<HashMap<CourseId, u8>> {
+    if groups_indices.len() as u8 == required_number_of_groups {
+        return check_if_completed_groups(sgs, groups_indices, courses);
+    }
+
```

Comment:
I think that all of the ```u8``` variables you defined in this file should be changed to ``` usize ```, to avoid the awkward 
```as u8``` conversion everywhere. this would make the code look cleaner, and the difference between u8 and usize in this context is really insignificant in terms of performance.